core/src/main/scala/org/http4s/ContextRequest.scala

```diff
@@ -0,0 +1,24 @@
+package org.http4s
+
+import cats.syntax.functor._
+import cats.Functor
+import cats.data.Kleisli
+
+final case class ContextRequest[F[_], A](context: A, req: Request[F])
+
+object ContextRequest {
+  def apply[F[_]: Functor, T](
+      getContext: Request[F] => F[T]): Kleisli[F, Request[F], ContextRequest[F, T]] =
+    Kleisli(request => getContext(request).map(ctx => ContextRequest(ctx, request)))
+
+
+  implicit class AuthInfo[F[_], A](req: ContextRequest[F, A]) {
```

Comment:
This feels awkward to me.  It's always going to be in scope for a `ContextRequest`, so why not just a member of `ContextRequest`?